### PR TITLE
[debug dump util] Route Module added

### PR DESCRIPTION
**Pull request by vivekreddynv**
Signed-off-by: Vivek Reddy Karri <vkarri@nvidia.com>

#### What I did
HLD for Dump Utility: [HLD](https://github.com/vivekreddynv/SONiC/pull/2).
* Added the Route Module to the Debug Dump Utility
* Added the Corresponding UT's

#### How to verify it

```
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_all_args PASSED                                                                                                                                                    [ 11%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_caching_redis_keys PASSED                                                                                                                                          [ 22%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_directly_connected_route PASSED                                                                                                                                    [ 33%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_ip2me_route PASSED                                                                                                                                                 [ 44%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_no_next_hop_id PASSED                                                                                                                                              [ 55%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_no_route_entry PASSED                                                                                                                                              [ 66%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_route_with_next_hop PASSED                                                                                                                                         [ 77%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_route_with_next_hop_group PASSED                                                                                                                                   [ 88%]
tests/dump_tests/module_tests/route_test.py::TestRouteModule::test_static_route PASSED   
```

#### New command output (if the output of a command-line utility has changed)

```
admin@r-tigon-11:~$ dump state route 200.0.1.0/26
{
    "200.0.1.0/26": {
        "APPL_DB": {
            "keys": [
                {
                    "ROUTE_TABLE:200.0.1.0/26": {
                        "ifname": "PortChannel0004,PortChannel0007",
                        "nexthop": "10.0.0.37,10.0.0.41"
                    }
                }
            ],
            "tables_not_found": []
        },
        "ASIC_DB": {
            "keys": [
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY:{\"dest\":\"200.0.1.0/26\",\"switch_id\":\"oid:0x21000000000000\",\"vr\":\"oid:0x3000000000002\"}": {
                        "SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID": "oid:0x5000000000c92"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_VIRTUAL_ROUTER:oid:0x3000000000002": {
                        "NULL": "NULL"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP:oid:0x5000000000c92": {
                        "SAI_NEXT_HOP_GROUP_ATTR_TYPE": "SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP:oid:0x5000000000c72": {
                        "SAI_NEXT_HOP_GROUP_ATTR_TYPE": "SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP:oid:0x5000000000c4f": {
                        "SAI_NEXT_HOP_GROUP_ATTR_TYPE": "SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP:oid:0x5000000000c57": {
                        "SAI_NEXT_HOP_GROUP_ATTR_TYPE": "SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER:oid:0x2d000000000c94": {
                        "SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID": "oid:0x5000000000c92",
                        "SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID": "oid:0x4000000000c49"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP_MEMBER:oid:0x2d000000000c93": {
                        "SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID": "oid:0x5000000000c92",
                        "SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID": "oid:0x4000000000c1c"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP:oid:0x4000000000c49": {
                        "SAI_NEXT_HOP_ATTR_IP": "10.0.0.41",
                        "SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID": "oid:0x6000000000b47",
                        "SAI_NEXT_HOP_ATTR_TYPE": "SAI_NEXT_HOP_TYPE_IP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP:oid:0x4000000000c1c": {
                        "SAI_NEXT_HOP_ATTR_IP": "10.0.0.37",
                        "SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID": "oid:0x6000000000b44",
                        "SAI_NEXT_HOP_ATTR_TYPE": "SAI_NEXT_HOP_TYPE_IP"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE:oid:0x6000000000b47": {
                        "SAI_ROUTER_INTERFACE_ATTR_MTU": "9100",
                        "SAI_ROUTER_INTERFACE_ATTR_PORT_ID": "oid:0x2000000000b1f",
                        "SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS": "1C:34:DA:1D:EA:00",
                        "SAI_ROUTER_INTERFACE_ATTR_TYPE": "SAI_ROUTER_INTERFACE_TYPE_PORT",
                        "SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID": "oid:0x3000000000002"
                    }
                },
                {
                    "ASIC_STATE:SAI_OBJECT_TYPE_ROUTER_INTERFACE:oid:0x6000000000b44": {
                        "SAI_ROUTER_INTERFACE_ATTR_MTU": "9100",
                        "SAI_ROUTER_INTERFACE_ATTR_PORT_ID": "oid:0x2000000000b1c",
                        "SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS": "1C:34:DA:1D:EA:00",
                        "SAI_ROUTER_INTERFACE_ATTR_TYPE": "SAI_ROUTER_INTERFACE_TYPE_PORT",
                        "SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID": "oid:0x3000000000002"
                    }
                }
            ],
            "tables_not_found": [],
            "vidtorid": {
                "oid:0x3000000000002": "oid:0x3",
                "oid:0x5000000000c92": "oid:0x3900000005",
                "oid:0x5000000000c72": "oid:0x3800000005",
                "oid:0x5000000000c4f": "oid:0x3300000005",
                "oid:0x5000000000c57": "oid:0x3500000005",
                "oid:0x2d000000000c94": "oid:0x3000390000002d",
                "oid:0x2d000000000c93": "oid:0x500390000002d",
                "oid:0x4000000000c49": "oid:0x3000000004",
                "oid:0x4000000000c1c": "oid:0x500000004",
                "oid:0x6000000000b47": "oid:0x1c100000006",
                "oid:0x6000000000b44": "oid:0x10100000006"
            }
        }
    }
}
```

### CODE REVIEW

**Review by dgsudharsan — DISMISSED** [review]

**Comment by dgsudharsan** [comment]
@prsunny @SuvarnaMeenakshi  Can you please review?

**Comment by rlhui** [comment]
@TACappleman - would you please review? Thanks.

File: dump/plugins/route.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        self.init_route_appl_info()
+        # ASIC DB - ROUTE ENTRY
+        self.nh_id, vr = self.init_asic_route_entry_info()
+        # ASIC DB - VIRTUAL ROUTER
+        self.init_asic_vr_info(vr)
+        # ASIC DB - KEYS dependent on NEXT HOP ID
+        self.init_asic_nh()
+        return self.ret_temp
+
+    def add_to_ret_template(self, table, db, keys, err, add_to_tables_not_found=True):
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/route.py:144]: There's another such refactoring request here: https://github.com/Azure/sonic-utilities/pull/1853#discussion_r746167894, 

I'll also do this after once the Route and VxLan modules are merged. It'll be cleaner that way

File: dump/plugins/route.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+    def execute(self, params):
+        self.ret_temp = create_template_dict(dbs=["CONFIG_DB", "APPL_DB", "ASIC_DB"])
+        self.dest_net = params[Route.ARG_NAME]
+        self.ns = params["namespace"]
+        # CONFIG DB
+        if not self.init_route_config_info():
+            del self.ret_temp["CONFIG_DB"]
+        # APPL DB
+        self.init_route_appl_info()
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/route.py]: Added

File: dump/plugins/route.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    2) SAI_OBJECT_TYPE_NEXT_HOP
+    3) SAI_OBJECT_TYPE_ROUTER_INTERFACE
+    Note: Caches all the kv pairs for a key
+    """
+
+    def __init__(self, m_engine):
+        self.key_cache = {}
+        self.m_engine = m_engine
+
+    def mutate_request(self, req):
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply dump/plugins/route.py]: Handled

**Comment by vivekreddynv** [comment]
@TACappleman, I've addressed the comments. Please review them

**Review by TACappleman — APPROVED** [review]

**Review by dgsudharsan — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Pull request merged**